oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,492 @@
+
+Reference Implementation
+************************
+
+A proof of concept has already been created demonstrating how Webpack module federation works with two Open edX micro-frontends based on the frontend-template-application. The POC has several shared libraries (``react``, ``react-dom``, and ``@openedx/paragon``), and loads a React component module from a guest MFE into the page of a host MFE. It supports hot module replacement during development, and runs on the two MFEs' `webpack-dev-server` instances.
+
+The POC repositories can be found here:
+
+- `frontend-app-mf-host <https://github.com/davidjoy/frontend-app-mf-host>`_
+- `frontend-app-mf-guest <https://github.com/davidjoy/frontend-app-mf-guest>`_
+
+Proposed MFE Architecture
+=========================
+
+.. image:: oep-0065/proposed-mfe-architecture.png
+
+MFEs and Modules
+----------------
+
+Each of our MFEs will export a set of one or more modules that can be loaded by other MFEs. For instance, ``frontend-app-profile`` would likely export the ``ProfilePage`` component. Other MFEs may export their own pages, or perhaps plugins/widgets/components to be loaded by the `frontend-plugin-framework <https://github.com/openedx/frontend-plugin-framework>`_ via a "module" plugin type based on this implementation.
+
+Hosts and Guests
+----------------
+
+MFEs become either "hosts" or "guests" or both. A host is an MFE that loads runtime modules from a guest. A guest may itself act as a host to modules from another guest.
+
+Shell MFE
+---------
+
+We will create a new "shell" MFE to act as the top-level host for all other MFEs. It is exclusively responsible for:
+
+- Initializing the application via ``@edx/frontend-platform``.
+- Loading the base, expected version of all our shared dependencies.
+- Rendering the "layout" of the application, including the header and footer.
+- Loading the brand.
+
+Like other hosts, it is also responsible for:
+
+- Loading all the manifests from the "guest" MFEs it intends to load.
+- Using module federation to load the guest MFEs on demand.
+
+Guest MFEs (not the shell)
+--------------------------
+
+Guest MFEs that require a version of a shared dependency that's incompatible with the shell's version may load their own, provided that dependency isn't a singleton (like ``react`` or ``@edx/frontend-platform``). Singletons may only be loaded once because they break if they're loaded more than once on a single page.
+
+Guests loading their own versions of shared dependencies degrades the performance and experience of end users. MFE authors should endeavor to use dependencies compatible with the version loaded by the shell. If we use a passthrough library of shared dependencies, this becomes easier.
+
+Converting the POC to a reference implementation
```

Comment:
This feels like it should definitely be a living document. There are robust comments, questions, and suggestions in the comments for each of the line items that shouldn't be lost - and discussion should continue, past when this OEP is merged.